src/adaptors/adaptor_common.c

```diff
@@ -72,24 +78,70 @@ qd_error_t qd_load_adaptor_config(qd_adaptor_config_t *config, qd_entity_t *enti
     return qd_error_code();
 }
 
+void qd_adaptor_common_init(void)
+{
+    if (buffer_ceiling > 0) {
+        return;
+    }
+
+    char     *ceiling_string = getenv("SKUPPER_MEMORY_CEILING");
+    uint64_t  memory_ceiling = (uint64_t) 4 * (uint64_t) 1024 * (uint64_t) 1024 * (uint64_t) 1024;  // 4 Gig default
+
+    if (!!ceiling_string) {
+        long long convert = atoll(ceiling_string);
+        if (convert > 0) {
+            memory_ceiling = (uint64_t) convert;
+        }
+    }
+
+    buffer_ceiling = MAX(memory_ceiling / QD_ADAPTOR_MAX_BUFFER_SIZE, 100);
+    buffer_threshold_50 = buffer_ceiling / 2;
+    buffer_threshold_75 = (buffer_ceiling / 20) * 15;
+    buffer_threshold_85 = (buffer_ceiling / 20) * 17;
+}
+
 int qd_raw_connection_grant_read_buffers(pn_raw_connection_t *pn_raw_conn)
 {
+#define TIER_1 8  // [0% .. 50%)
+#define TIER_2 4  // [50% .. 75%)
+#define TIER_3 2  // [75% .. 85%)
+#define TIER_4 2  // [85% .. 100%]
+
     assert(pn_raw_conn);
     pn_raw_buffer_t raw_buffers[RAW_BUFFER_BATCH];
-    size_t          desired = pn_raw_connection_read_buffers_capacity(pn_raw_conn);
-    const size_t    granted = desired;
+    size_t desired = TIER_4;
+    size_t capacity = MIN(pn_raw_connection_read_buffers_capacity(pn_raw_conn), TIER_1);
```

Comment:
There may be a problem with this that results in granting more buffers than desired. (or I may be missing something obvious).

If this function is called multiple times before proton actually consumes the buffers I think that the raw connection will be filled to the maximum supported by the raw connection (16 currently) regardless of the memory pressure.

In the TCP adaptor this routine is called every time the delivery continue event is handled.  I've tried setting SKUPPER_MEMORY_CEILING to 65536 which should be hit at 4 buffers, but when I run the CI test it appears as if the raw connection is always getting 16 read buffers anyway.

The problem is we don't have an API to query how many rx bufs the raw connection already has on entry.  We could assume 16 since that's the current raw connection impl.

Reply:
@kgiusti I believe you are correct here.  To do this right, we need to know either the number of granted read buffers or the maximum number of read buffers.

If we set TIER_1 to 16, I think it would work correctly by virtue of the coincidence that TIER_1 == max_buffers.

Reply:
This is now fixed.